cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/AvroFileManager.java

```diff
@@ -0,0 +1,248 @@
+  private final Map<LogPathIdentifier, LogFileOutputStream> fileMap;
+  private final Location logsDirectoryLocation;
+  private final FileMetaDataManager fileMetaDataManager;
+  private final int syncIntervalBytes;
+  private final Schema schema;
+
+  AvroFileManager(long maxFileLifetimeMs, int syncIntervalBytes, Schema schema,
+                  FileMetaDataManager fileMetaDataManager,
+                  LocationFactory locationFactory) {
+    this.maxLifetimeMillis = maxFileLifetimeMs;
+    this.syncIntervalBytes = syncIntervalBytes;
+    this.schema = schema;
+    this.fileMetaDataManager = fileMetaDataManager;
+    this.logsDirectoryLocation = locationFactory.create("logs");
+    this.fileMap = new HashMap<>();
+  }
+
+  /**
+   * Get an avro file for the give logging context and timestamp - return an already open file,
+   * or create and return a new file or rotate a file based on time and return.
+   * @param logPathIdentifier
+   * @param timestamp
+   * @return LogFileOutputStream
+   * @throws IOException
+   */
+  public LogFileOutputStream getAvroFile(LogPathIdentifier logPathIdentifier, long timestamp) throws IOException {
+    LogFileOutputStream logFileOutputStream = fileMap.get(logPathIdentifier);
+    // If the file is not open then set reference to null so that a new one gets created
+    if (logFileOutputStream == null) {
+      logFileOutputStream = createAvroFile(logPathIdentifier, timestamp);
+    }
+    // rotate the file if needed (time has passed)
+    logFileOutputStream = rotateFile(logFileOutputStream, logPathIdentifier, timestamp);
+    return logFileOutputStream;
+  }
+
+  private LogFileOutputStream createAvroFile(LogPathIdentifier identifier, long timestamp) throws IOException {
+    Location location = getLocation(identifier, timestamp);
+    LOG.info("Creating Avro file {}", location);
+    LogFileOutputStream logFileOutputStream = new LogFileOutputStream(location);
+    try {
+      fileMetaDataManager.writeMetaData(identifier, timestamp, location);
+    } catch (Throwable e) {
+      closeAndDelete(logFileOutputStream);
+      throw new IOException(e);
+    }
+
+    fileMap.put(identifier, logFileOutputStream);
+    return logFileOutputStream;
+  }
+
+  private LogFileOutputStream rotateFile(LogFileOutputStream logFileOutputStream,
+                                         LogPathIdentifier identifier, long timestamp) throws IOException {
+    long currentTs = System.currentTimeMillis();
+    long timeSinceFileCreate = currentTs - logFileOutputStream.getCreateTime();
+    if (timeSinceFileCreate > maxLifetimeMillis) {
+      logFileOutputStream.close();
+      return createAvroFile(identifier, timestamp);
+    }
+    return logFileOutputStream;
+  }
+
+  /**
+   * closes all the open avro files in the map
+   */
+  public void close() {
+    // close all the files in fileMap
+    // clear the map
+    for (LogFileOutputStream file : fileMap.values()) {
+      try {
+        file.flush();
```

Comment:
Shouldn't need to call flush. OutputStream should have the contract that upon closing, it should flush everything.